lib/ansible/modules/windows/win_shortcut.ps1

```diff
@@ -0,0 +1,147 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# WANT_JSON
+# POWERSHELL_COMMON
+#
+# Based on: http://powershellblogger.com/2016/01/create-shortcuts-lnk-or-url-files-with-powershell/
+
+$ErrorActionPreference = "Stop"
+
+$params = Parse-Args $args -supports_check_mode $true;
+
+$_ansible_check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -default $false
+
+$orig_src = Get-AnsibleParam -obj $params -name "src" -default $null
+$orig_dest = Get-AnsibleParam -obj $params -name "dest" -failifempty $true
+$state = Get-AnsibleParam -obj $params -name "state" -default "present"
+$orig_args = Get-AnsibleParam -obj $params -name "args" -default $null
+$orig_directory = Get-AnsibleParam -obj $params -name "directory" -default $null
+$hotkey = Get-AnsibleParam -obj $params -name "hotkey" -default $null
+$orig_icon = Get-AnsibleParam -obj $params -name "icon" -default $null
+$orig_description = Get-AnsibleParam -obj $params -name "description" -default $null
+$windowstyle = Get-AnsibleParam -obj $params -name "windowstyle" -default $null
+
+# Expand environment variables (Beware: turns $null into "")
+$src = [System.Environment]::ExpandEnvironmentVariables($orig_src)
+$dest = [System.Environment]::ExpandEnvironmentVariables($orig_dest)
+$args = [System.Environment]::ExpandEnvironmentVariables($orig_args)
+$directory = [System.Environment]::ExpandEnvironmentVariables($orig_directory)
+$icon = [System.Environment]::ExpandEnvironmentVariables($orig_icon)
+$description = [System.Environment]::ExpandEnvironmentVariables($orig_description)
+
+$result = New-Object PSObject @{
+    changed = $false
+    dest = $dest
+    state = $state
+}
+
+If ($state -eq "absent") {
+    If (Test-Path "$dest") {
+        # If the shortcut exists, try to remove it
+        Try {
+            Remove-Item -Path "$dest";
+        } Catch {
+            # Report removal failure
+            Fail-Json $result "Failed to remove shortcut $dest. ($($_.Exception.Message))"
+        }
+        # Report removal success
+        Set-Attr $result "changed" $true
+    } Else {
+        # Nothing to report, everything is fine already
+    }
+} ElseIf ($state -eq "present") {
+    $Shell = New-Object -ComObject ("WScript.Shell")
+    $ShortCut = $Shell.CreateShortcut($dest)
+
+    # Compare existing values with new values,
+    # report as changed if required
+
+    If ($orig_src -ne $null -and $ShortCut.TargetPath -ne $src) {
+        Set-Attr $result "changed" $true
+        Set-Attr $result "src" $src
+        $ShortCut.TargetPath = $src
+    } Else {
+        Set-Attr $result "src" $ShortCut.TargetPath
+    }
+
+    # TODO: Find a better way to do has_attr() or isinstance() ?
+    If (Get-Member -InputObject $ShortCut -Name Arguments) {
```

Comment:
For COM/IDispatch objects like this without a managed interface wrapper, this is unfortunately about the best you can do (otherwise you'd just do `-is [type]`). The "technically proper" thing would be a `Marshal.QueryInterface` for the IID of WshShortcut/WshUrlShortcut, but that's pretty fugly. Given how ancient this stuff is and how unlikely it is that `Arguments` will magically appear on the WshUrlShortcut down the line, this is probably the best balance of readability/pedantic correctness. A comment explaining why you're doing it for the next guy might be good though, eg, "determine if we have a WshShortcut or WshUrlShortcut by checking for presence of Arguments property"...